050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,876 @@
+      size: 100Gi
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: brokers
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  roles
+    - broker
+  replicas: 6
+  resources:
+    requests:
+      memory: 64Gi
+      cpu: 8
+    limits:
+      memory: 64Gi
+      cpu: 12
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+    - id: 1
+      type: persistent-claim
+      size: 1000Gi
+```
+
+The KRaft mode will be supported only through the `KafkaNodePool` resources.
+
+### Handling of node IDs
+
+Node IDs (broker IDs) uniquely identify each of the Kafka cluster nodes.
+Node IDs will be assigned sequentially:
+
+* They will start with 0.
+* By default, when a new node ID is needed, the lowest possible available number will be used regardless the pool where the node will be added:
+    * In most cases, the node ID will be assigned from 0 up: 0, 1, 2, 3.
+    * If any gap will exist in the existing node IDs, it will be filled in.
+      For example when the existing nodes have IDs 0, 1, 4, 5, 6; the next node will get ID 2.
+* By default, when scaling down a pool, a highest used node ID will be removed.
+    * For example in Kafka cluster with two pools, where the first pool has IDs 0, 1 and 2 and the second pool has IDs 3, 4 and 5, when you scale down the first pool, the node with ID 2 will be removed and when you scale down the second pool node 5 will be removed.
+* The pods will reflect the node IDs in their name in the same way as they do today.
+
+The fact that the node IDs will not be assigned to the pools means the Pods will be mixed.
+In a cluster with two pools, they might look like this:
+
+* `my-cluster-big-nodes-0`
+* `my-cluster-big-nodes-1`
+* `my-cluster-small-nodes-2`
+* `my-cluster-small-nodes-3`
+* `my-cluster-big-nodes-4`
+* `my-cluster-small-nodes-5`
+
+This is anyway unavoidable if we want to allow nodes to move between the pools.
+
+To allow the operator to keep track of the assigned node IDs, it will keep track of them in the `.status` section of the `KafkaNodePool` resource.
+That way, it will be able to know at any point in time which IDs are used and which are available.
+It will be also able to identify the scale-down or scale-up events using this information.
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: big-nodes
+  # ...
+spec:
+  replicas: 3
+  # ...
+status:
+  nodeIds:
+    - 0
+    - 1
+    - 2
+  clusterId: cABpUKeDTqi7q_XOtBMgUv
+  # replicas and labelSelector fields are needed for the scale subresource
+  replicas: 3
+  labelSelector: >-
+    strimzi.io/cluster=my-cluster,strimzi.io/name=m-cluster-big-nodes,strimzi.io/kind=Kafka
+```
+
+Users will be able to use annotations to indicate to the operator what should be the next ID when scaling up or what should be the node ID which should be removed when scaling down through annotations.
```

Comment:
Which resources will these annotations be on? I assume it's the node pools themselves, but it's best to be explicit.